tsl/src/compression/compression.c

```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
 		CommandCounterIncrement();
 	}
 }
+
+/* Returns a Bitmap with all columns part of unique constraints */
+static Bitmapset *
+relation_unique_columns(Relation rel)
+{
+	/*
+	 * We are using the list of indexes to determine columns
+	 * part of unique constraints here because the index list
+	 * is cached. It might be more correct to use the constraints
+	 * of the relation but getting those requires a sequential
+	 * scan on the catalog.
+	 */
+	List *indexoidlist = RelationGetIndexList(rel);
+	ListCell *lc;
+
+	/* Bitmapset of unique columns in the uncompressed chunk */
+	Bitmapset *unique_columns = NULL;
+
+	foreach (lc, indexoidlist)
+	{
+		Oid indexoid = lfirst_oid(lc);
+		HeapTuple index_tuple;
+		Form_pg_index index;
+
+		index_tuple = SearchSysCache1(INDEXRELID, ObjectIdGetDatum(indexoid));
+		if (!HeapTupleIsValid(index_tuple))
+			elog(ERROR,
+				 "cache lookup failed for index %u in \"%s\" ",
+				 indexoid,
+				 RelationGetRelationName(rel));
+		index = (Form_pg_index) GETSTRUCT(index_tuple);
+
+		if (index->indisunique)
+		{
+			for (int i = 0; i < index->indnkeyatts; i++)
+			{
+				unique_columns = bms_add_member(unique_columns, index->indkey.values[i]);
+			}
+		}
+
+		ReleaseSysCache(index_tuple);
+	}
+
+	list_free(indexoidlist);
+
+	return unique_columns;
+}
+
+static ScanKeyData *
+build_scankeys(int32 hypertable_id, RowDecompressor decompressor, Bitmapset *unique_columns,
```

Comment:
This part looks like "planning-time", similar to the deduction of arbiter indexes. But probably there's no convenient way to do it there. I think it would be good not do it per-row and cache this in the chunk insert state.

Reply:
This needs to be per row as the scankeys are built from row values. I dont think you can easily cache this.

Reply:
Postgres index scans do that by just replacing the `sk_argument` (`ExecIndexEvalRuntimeKeys`).

Reply:
i'll leave that for a followup PR